File: Emulators/EmulatorManager/PCSX2Manager.cs
using PointerFinder2.Core;
using PointerFinder2.DataModels;
using System;
using System.Diagnostics;
using System.Linq;

namespace PointerFinder2.Emulators.EmulatorManager
{
    public class Pcsx2Manager : IEmulatorManager
    {
        private readonly DebugLogForm logger = DebugLogForm.Instance;

        // PS2 Memory Layout Constants.
        public const uint PS2_EEMEM_START = 0x20000000;
        public const uint PS2_EEMEM_END = 0x21FFFFFF + 1; // 32MB of EE RAM
        public const uint PS2_GAME_CODE_START = 0x00100000;
        public const uint PS2_GAME_CODE_END = 0x01FFFFFF + 1; // Main game code region

        #region Interface Implementation
        public string EmulatorName => "PCSX2";
        public uint MainMemoryStart => PS2_EEMEM_START;
        public uint MainMemorySize => 32 * 1024 * 1024; // 32MB
        public string RetroAchievementsPrefix => "X";
        public Process EmulatorProcess { get; private set; }
        public bool IsAttached => ProcessHandle != IntPtr.Zero && MemoryBasePC != IntPtr.Zero;
        public nint ProcessHandle { get; private set; } = IntPtr.Zero;
        public nint MemoryBasePC { get; private set; } = IntPtr.Zero;
        private nint Ps2MemoryBaseInPC { get; set; } = IntPtr.Zero;

        // Attaches by finding PCSX2's exported "EEMem" variable, which points to the emulated RAM.
        public bool Attach(Process process)
        {
            if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] Attempting to attach...");

            EmulatorProcess = process;

            if (EmulatorProcess?.MainModule == null)
            {
                if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] FAILURE: Process not found or main module is not accessible.");
                return false;
            }
            if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] SUCCESS: Attaching to process '{EmulatorProcess.ProcessName}' (ID: {EmulatorProcess.Id}).");

            ProcessHandle = Memory.OpenProcessHandle(EmulatorProcess);
            if (ProcessHandle == IntPtr.Zero)
            {
                if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] FAILURE: Could not open process handle. Try running this tool as Administrator.");
                return false;
            }
            if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] SUCCESS: Process handle opened.");

            nint eeMemBasePC = Memory.FindExportedAddress(EmulatorProcess, ProcessHandle, "EEMem");
            if (eeMemBasePC == IntPtr.Zero)
            {
                if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] FAILURE: Could not find the 'EEMem' export. Ensure the game is fully loaded.");
                return false;
            }

            long? eeMemPtr = Memory.ReadInt64(ProcessHandle, eeMemBasePC);
            if (!eeMemPtr.HasValue)
            {
                if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] FAILURE: Could not read the pointer from the 'EEMem' export address.");
                return false;
            }
            MemoryBasePC = (nint)eeMemPtr.Value;
            if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] SUCCESS: EEMem base found in PC memory at 0x{MemoryBasePC:X}.");

            Ps2MemoryBaseInPC = IntPtr.Subtract(MemoryBasePC, (int)PS2_EEMEM_START);
            if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] Attachment complete. Ready for scanning.");
            return true;
        }

        // Detaches from the emulator process and cleans up resources.
        public void Detach()
        {
            if (ProcessHandle != IntPtr.Zero)
            {
                Memory.CloseHandle(ProcessHandle);
            }
            ProcessHandle = IntPtr.Zero;
            MemoryBasePC = IntPtr.Zero;
            Ps2MemoryBaseInPC = IntPtr.Zero;
            EmulatorProcess = null;
            if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] Detached from process.");
        }

        // Reads a block of memory from the emulated PS2 RAM.
        public byte[] ReadMemory(uint ps2Address, int count)
        {
            if (!IsAttached) return null;
            nint address = IntPtr.Add(Ps2MemoryBaseInPC, (int)ps2Address);
            return Memory.ReadBytes(ProcessHandle, address, count);
        }

        // Reads a 32-bit unsigned integer from a PS2 memory address.
        public uint? ReadUInt32(uint ps2Address)
        {
            byte[] buffer = ReadMemory(ps2Address, 4);
            if (buffer == null)
            {
                return null;
            }
            return BitConverter.ToUInt32(buffer, 0);
        }

        // Checks if a value is a valid pointer target within the PS2's main memory regions.
        public bool IsValidPointerTarget(uint value)
        {
            if ((value & 3) != 0) return false; // Must be 4-byte aligned.
            bool isGameCode = value >= PS2_GAME_CODE_START && value < PS2_GAME_CODE_END;
            bool isEeMem = value >= PS2_EEMEM_START && value < PS2_EEMEM_END;
            return isGameCode || isEeMem;
        }

        // Traverses a pointer path to verify its final address. Used for filtering.
        public uint? RecalculateFinalAddress(PointerPath path, uint expectedFinalAddress)
        {
            bool shouldLog = DebugSettings.LogFilterValidation;

            (expectedFinalAddress, _) = NormalizeAddressForRead(expectedFinalAddress);

            if (shouldLog)
            {
                logger.Log("--------------------------------------------------");
                logger.Log($"[{EmulatorName}] Validating Path: {path.BaseAddress:X8} -> {path.GetOffsetsString()} (Expecting: 0x{expectedFinalAddress:X8})");
            }

            if (path == null || !path.Offsets.Any())
            {
                if (shouldLog) logger.Log($"[{EmulatorName}] FAILURE: Path is null or has no offsets.");
                return null;
            }

            if (shouldLog) logger.Log($"[{EmulatorName}] Step 0: Reading base address [0x{path.BaseAddress:X8}]");
            uint? currentAddress = ReadUInt32(path.BaseAddress);
            if (!currentAddress.HasValue)
            {
                if (shouldLog) logger.Log($"[{EmulatorName}] FAILURE: Could not read base address.");
                return null;
            }
            if (!IsValidPointerTarget(currentAddress.Value))
            {
                if (shouldLog) logger.Log($"[{EmulatorName}] FAILURE: Base address points to invalid memory 0x{currentAddress.Value:X8}.");
                return null;
            }
            if (shouldLog) logger.Log($"  -> Value: 0x{currentAddress.Value:X8}");

            for (int i = 0; i < path.Offsets.Count - 1; i++)
            {
                if (shouldLog) logger.Log($"[{EmulatorName}] Step {i + 1}: Applying Offset #{i + 1} ({path.Offsets[i]:+X;-X})");

                uint nextAddressToRead = currentAddress.Value + (uint)path.Offsets[i];
                if (shouldLog) logger.Log($"  -> Calculating next address to read: 0x{currentAddress.Value:X8} + 0x{path.Offsets[i]:X} = 0x{nextAddressToRead:X8}");

                (nextAddressToRead, bool wasNorm) = NormalizeAddressForRead(nextAddressToRead);
                if (wasNorm && shouldLog) logger.Log($"  -> Normalizing address to 0x{nextAddressToRead:X8}");

                if (shouldLog) logger.Log($"  -> Reading from [0x{nextAddressToRead:X8}]");
                currentAddress = ReadUInt32(nextAddressToRead);

                if (!currentAddress.HasValue)
                {
                    if (shouldLog) logger.Log($"[{EmulatorName}] FAILURE: Could not read memory at 0x{nextAddressToRead:X8}. Chain is broken.");
                    return null;
                }
                if (!IsValidPointerTarget(currentAddress.Value))
                {
                    if (shouldLog) logger.Log($"[{EmulatorName}] FAILURE: Intermediate pointer is invalid memory 0x{currentAddress.Value:X8}.");
                    return null;
                }
                if (shouldLog) logger.Log($"  -> Value: 0x{currentAddress.Value:X8}");
            }

            if (shouldLog) logger.Log($"[{EmulatorName}] Step {path.Offsets.Count}: Applying Final Offset ({path.Offsets.Last():+X;-X})");
            uint finalAddress = currentAddress.Value + (uint)path.Offsets.Last();
            if (shouldLog) logger.Log($"  -> Final Calculation: 0x{currentAddress.Value:X8} + 0x{path.Offsets.Last():X} = 0x{finalAddress:X8}");

            (finalAddress, bool finalNorm) = NormalizeAddressForRead(finalAddress);
            if (finalNorm && shouldLog) logger.Log($"  -> Normalizing FINAL address to 0x{finalAddress:X8}");

            if (shouldLog)
            {
                bool isValid = finalAddress == expectedFinalAddress;
                logger.Log($"[{EmulatorName}] Comparison: Calculated (0x{finalAddress:X8}) == Expected (0x{expectedFinalAddress:X8}) -> {isValid.ToString().ToUpper()}");
            }

            return finalAddress;
        }

        // Formats a full PS2 address into a shorter format relative to EE RAM for display.
        public string FormatDisplayAddress(uint address)
        {
            if (address >= 0x20000000 && address < 0x22000000)
            {
                return (address - 0x20000000).ToString("X");
            }
            return address.ToString("X8");
        }

        // Converts a user-entered address into a full PS2 memory address.
        public uint UnnormalizeAddress(string address)
        {
            uint parsedAddress = uint.Parse(address.Replace("0x", ""), System.Globalization.NumberStyles.HexNumber);
            if (parsedAddress < 0x2000000)
            {
                return parsedAddress + 0x20000000;
            }
            return parsedAddress;
        }

        // Handles PS2's address mirroring (e.g., 0x0... -> 0x20...).
        public (uint normalizedAddress, bool wasNormalized) NormalizeAddressForRead(uint address)
        {
            if (address < PS2_EEMEM_START)
            {
                return (address + PS2_EEMEM_START, true);
            }
            return (address, false);
        }

        // Compares two PS2 addresses, accounting for mirroring.
        public bool AreAddressesEquivalent(uint addr1, uint addr2)
        {
            uint normAddr1 = (addr1 < PS2_EEMEM_START) ? addr1 + PS2_EEMEM_START : addr1;
            uint normAddr2 = (addr2 < PS2_EEMEM_START) ? addr2 + PS2_EEMEM_START : addr2;
            return normAddr1 == normAddr2;
        }

        // Provides a set of default settings specifically for PCSX2.
        public AppSettings GetDefaultSettings()
        {
            if (DebugSettings.LogLiveScan) logger.Log($"[{EmulatorName}] Getting default settings.");
            return new AppSettings
            {
                StaticAddressStart = "100000",
                StaticAddressEnd = "7FFFFF",
                MaxOffset = 4095,
                MaxLevel = 7,
                MaxResults = 500000,
                // Provide default for new MaxCandidates setting.
                MaxCandidates = 10000000,
                ScanForStructureBase = false,
                MaxNegativeOffset = 1024,
                Use16ByteAlignment = true,
                StopOnFirstPathFound = false,
                CandidatesPerLevel = 10
            };
        }

        // Added implementation for the new interface method.
        public long GetIndexForStateDump(uint address)
        {
            (uint normalizedAddress, _) = NormalizeAddressForRead(address);
            if (normalizedAddress >= MainMemoryStart && normalizedAddress < (MainMemoryStart + MainMemorySize))
            {
                return normalizedAddress - MainMemoryStart;
            }
            return -1;
        }
        #endregion
    }
}

File: src/Emulators/Managers/Pcsx2Manager.cpp
#include "Pcsx2Manager.h"

namespace PointerFinder2::Emulators::EmulatorManager {

    Pcsx2Manager::Pcsx2Manager() = default;

    QString Pcsx2Manager::getEmulatorName() const {
        return "PCSX2";
    }

    uint32_t Pcsx2Manager::getMainMemoryStart() const {
        return PS2_EEMEM_START;
    }

    uint32_t Pcsx2Manager::getMainMemorySize() const {
        return 32 * 1024 * 1024; // 32MB Emotion Engine RAM
    }

    QString Pcsx2Manager::getRetroAchievementsPrefix() const {
        return "X";
    }

    bool Pcsx2Manager::attach(uint32_t processId) {
        detach();
        m_processId = processId;
        m_processHandle = Core::Memory::openProcessHandle(m_processId);

        if (!m_processHandle) {
            return false;
        }

        uintptr_t baseAddress = Core::Memory::getModuleBaseAddress(m_processHandle);
        if (baseAddress == 0) {
            detach();
            return false;
        }

        // Resolves primary EE segment references exported by the core execution module.
        uintptr_t eeMemBasePC = Core::Memory::findExportedAddress(m_processHandle, baseAddress, "EEMem", true);
        if (eeMemBasePC == 0) {
            eeMemBasePC = Core::Memory::findExportedAddress(m_processHandle, baseAddress, "EEMem", false);
        }

        if (eeMemBasePC == 0) {
            detach();
            return false;
        }

        auto eeMemPtr = Core::Memory::readInt64(m_processHandle, eeMemBasePC);
        if (!eeMemPtr.has_value()) {
            detach();
            return false;
        }

        m_memoryBasePC = static_cast<uintptr_t>(eeMemPtr.value());
        m_ps2MemoryBaseInPC = m_memoryBasePC - PS2_EEMEM_START;
        return true;
    }

    // Verifies if the PCSX2 'EEMem' connector handle is still pointing to active PS2 memory.
    bool Pcsx2Manager::verifyAttachment() {
        if (!isAttached()) return false;

        uintptr_t baseAddress = Core::Memory::getModuleBaseAddress(m_processHandle);
        if (baseAddress == 0) return false;

        uintptr_t eeMemBasePC = Core::Memory::findExportedAddress(m_processHandle, baseAddress, "EEMem", true);
        if (eeMemBasePC == 0) {
            eeMemBasePC = Core::Memory::findExportedAddress(m_processHandle, baseAddress, "EEMem", false);
        }
        if (eeMemBasePC == 0) return false;

        auto eeMemPtr = Core::Memory::readInt64(m_processHandle, eeMemBasePC);
        if (!eeMemPtr.has_value()) return false;

        return m_memoryBasePC == static_cast<uintptr_t>(eeMemPtr.value());
    }

    // Reads raw data directly from the emulator's active Emotion Engine memory bank.
    std::vector<uint8_t> Pcsx2Manager::readMemory(uint32_t address, size_t count) {
        if (!isAttached()) return {};
        uintptr_t targetPC = m_ps2MemoryBaseInPC + address;
        return Core::Memory::readBytes(m_processHandle, targetPC, count);
    }

    // Checks if a given value is a valid address inside standard PS2 memory blocks.
    bool Pcsx2Manager::isValidPointerTarget(uint32_t value) const {
        if ((value & 3) != 0) return false;
        bool isGameCode = value >= PS2_GAME_CODE_START && value < PS2_GAME_CODE_END;
        bool isEeMem = value >= PS2_EEMEM_START && value < PS2_EEMEM_END;
        return isGameCode || isEeMem;
    }

    bool Pcsx2Manager::isPotentialPointer(uint32_t value) const {
        bool isEemem = (value >= PS2_EEMEM_START && value < PS2_EEMEM_END);
        bool isGameCode = (value >= PS2_GAME_CODE_START && value < PS2_GAME_CODE_END) && ((value & 0xF) == 0);
        return isEemem || isGameCode;
    }

    bool Pcsx2Manager::isValidPlatformPointer(uint32_t value) const {
        return isPotentialPointer(value) && ((value & 0xF) == 0);
    }

    // Formats raw addresses into clean strings, omitting standard PS2 system prefixes.
    QString Pcsx2Manager::formatDisplayAddress(uint32_t address) const {
        if (address >= PS2_EEMEM_START && address < PS2_EEMEM_END) {
            return QString::number(address - PS2_EEMEM_START, 16).toUpper();
        }
        return QString::number(address, 16).toUpper().rightJustified(8, '0');
    }

    // Translates a plain user hex string back into a system-specific PS2 memory address.
    uint32_t Pcsx2Manager::unnormalizeAddress(const QString& address) const {
        bool ok;
        uint32_t parsed = address.toUInt(&ok, 16);
        if (!ok) return 0;

        if (parsed < 0x2000000) {
            return parsed + PS2_EEMEM_START;
        }
        return parsed;
    }

    // Restores missing PS2 prefixes (0x20000000) to addresses so they can be read correctly.
    std::pair<uint32_t, bool> Pcsx2Manager::normalizeAddressForRead(uint32_t address) const {
        if (address < PS2_EEMEM_START) {
            return { address + PS2_EEMEM_START, true };
        }
        return { address, false };
    }

    bool Pcsx2Manager::areAddressesEquivalent(uint32_t addr1, uint32_t addr2) const {
        uint32_t norm1 = (addr1 < PS2_EEMEM_START) ? addr1 + PS2_EEMEM_START : addr1;
        uint32_t norm2 = (addr2 < PS2_EEMEM_START) ? addr2 + PS2_EEMEM_START : addr2;
        return norm1 == norm2;
    }

    DataModels::AppSettings Pcsx2Manager::getDefaultSettings() const {
        DataModels::AppSettings settings;
        settings.staticAddressStart = "100000";
        settings.staticAddressEnd = "7FFFFF";
        settings.maxOffset = 4095;
        settings.maxLevel = 7;
        settings.maxCandidates = 10000000;
        settings.stopOnFirstPathFound = false;
        settings.candidatesPerLevel = 10;
        settings.fastScanMode = true;
        return settings;
    }

    // Translates an active system address into a physical offset index in the captured snapshot file.
    int64_t Pcsx2Manager::getIndexForStateDump(uint32_t address) const {
        auto [normed, wasNorm] = normalizeAddressForRead(address);
        if (normed >= PS2_EEMEM_START && normed < PS2_EEMEM_END) {
            return normed - PS2_EEMEM_START;
        }
        return -1;
    }

}